docs/src/tutorials/conic/ellipse_fitting.jl

```diff
@@ -0,0 +1,416 @@
+# discrete wavelet transform (DWT) in combination with the Iterative Shrinking and 
+# Thresholding (ISTA) algorithm to denoise the image and find a sparse representation.
+# This will remove redundant information and make it much easier to detect the edges of
+# galaxies.
+
+# Finding a sparse representation amounts to solving the following optimization problem:
+# ```math
+# \begin{equation}
+#     \min_{x} \frac{1}{2} \left\| y - \Phi x \right\|_2^2 + \lambda \left\| x \right\|_1
+# \end{equation}
+# ```
+# where ``y`` is the noisy image, ``\Phi`` is the sparsifying basis, ``x`` is the sparse
+# representation of our image, and ``\lambda`` is the regularization parameter which 
+# we set to 0.1.
+
+# To work with our image we must first convert it to Float64.
+x = convert(Array{Float64}, img_gray_roi)
+
+# We then use ISTA in combination with our wavelet sparsifying basis ``\Psi`` obtained 
+# from the [family of Daubechies wavelets](https://en.wikipedia.org/wiki/Daubechies_wavelet). 
+# We use the db4 wavelet which has 4 vanishing moments. We set the number of iterations 
+# to 15.
+reg = L1Regularization(0.1);
+Φ = WaveletOp(Float64; shape = size(x), wt = wavelet(WT.db4));
+solver = createLinearSolver(OptISTA, Φ; reg = reg, iterations = 15);
+
+# The sampled image in wavelet domain is given by:
+b = Φ * vec(x);
+
+# We can now solve the optimization problem to find the sparse representation of the image.
+x_approx = solve!(solver, b)
+x_approx = reshape(x_approx, size(x));
+x_final = normalize(x_approx)
+
+mosaicview(x, Gray.(x_final); nrow = 1)
+
+# We then use a binarization algorithm to map each grayscale pixel 
+# ``(x_i, y_i)`` to a binary value so ``x_i, y_i \to \{0, 1\}``.
+x_bin = binarize(x_final, Otsu(); nbins = 128)
+x_bin = convert(Array{Bool}, x_bin)
+
+plt = plot_dwt(img_roi, (500, 500))
+heatmap!(x_bin; color = :grays, alpha = 0.45)
+
+# ## Edge detection and clustering
+
+# Now that we have our binary image, we can use edge detection to find the edges of the
+# galaxies. We will use the [Sobel operator](https://en.wikipedia.org/wiki/Sobel_operator) for this task.
+# <details>
+# <summary>Sobel operator</summary>
+function edge_detector(
+    f_smooth::Matrix{Float64},
+    d1::Float64 = 0.1,
+    d2::Float64 = 0.1,
+)
+    rows, cols = size(f_smooth)
+
+    gradient_magnitude = zeros(Float64, rows, cols)
+    laplacian_magnitude = zeros(Float64, rows, cols)
+    edge_matrix = zeros(Bool, rows, cols)  # Binary matrix for edges
+
+    sobel_x = [-1 0 1; -2 0 2; -1 0 1]
+    sobel_y = [-1 -2 -1; 0 0 0; 1 2 1]
+    sobel_xx = [-1 2 -1; 2 -4 2; -1 2 -1]
+    sobel_yy = [-1 2 -1; 2 -4 2; -1 2 -1]
+
+    gradient_x = conv(f_smooth, sobel_x)
+    gradient_y = conv(f_smooth, sobel_y)
+
+    gradient_magnitude = sqrt.(gradient_x .^ 2 + gradient_y .^ 2)
+
+    gradient_xx = conv(f_smooth, sobel_xx)
+    gradient_yy = conv(f_smooth, sobel_yy)
+    laplacian_magnitude = sqrt.(gradient_xx .^ 2 + gradient_yy .^ 2)
+
+    for i in 1:rows
```

Comment:
```suggestion
    for i in 1:rows, j in 1:cols
```
You can drop one level of indentation with this.

Or just
```julia
    return (gradient_magnitude .> d1) .| (laplacian_magnitude .< d2) 
```